DigitalLearningSolutions.Data.Tests/Services/DelegateDownloadFileServiceTests.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            using var expectedWorkbook = new XLWorkbook(
+                TestContext.CurrentContext.TestDirectory + "\\TestData\\DelegateUploadTest.xlsx"
+            );
+
+            A.CallTo(() => jobGroupsDataService.GetJobGroupsAlphabetical()).Returns(
+                new[] { (2, "Doctor"), (3, "Health Professional"), (1, "Nursing") }
+            );
+
+            A.CallTo(() => userDataService.GetDelegateUserCardsByCentreId(2)).Returns(
+                new List<DelegateUserCard>
+                {
+                    new DelegateUserCard
+                    {
+                        FirstName = "A",
+                        LastName = "Test",
+                        EmailAddress = null,
+                        CandidateNumber = "TT95",
+                        Answer1 = "xxxx",
+                        Answer2 = "xxxxxxxxx",
+                        Answer3 = null,
+                        Answer4 = null,
+                        Answer5 = null,
+                        Answer6 = null,
+                        Active = true,
+                        AliasId = null,
+                        JobGroupId = 1
+                    },
+                    new DelegateUserCard
+                    {
+                        FirstName = "Fake",
+                        LastName = "Person",
+                        EmailAddress = "Test@Test",
+                        CandidateNumber = "TU67",
+                        Answer1 = null,
+                        Answer2 = null,
+                        Answer3 = null,
+                        Answer4 = null,
+                        Answer5 = null,
+                        Answer6 = null,
+                        Active = true,
+                        AliasId = null,
+                        JobGroupId = 1
+                    },
+                    new DelegateUserCard
+                    {
+                        FirstName = "Test",
+                        LastName = "User",
+                        EmailAddress = null,
+                        CandidateNumber = "TM323",
+                        Answer1 = "xx",
+                        Answer2 = "xxxxxxxx",
+                        Answer3 = null,
+                        Answer4 = null,
+                        Answer5 = null,
+                        Answer6 = null,
+                        Active = true,
+                        AliasId = null,
+                        JobGroupId = 2
+                    }
+                }
+            );
+
+            // When
+            var resultBytes = delegateDownloadFileService.GetDelegateDownloadFileForCentre(2);
+            using var resultsStream = new MemoryStream(resultBytes);
+            using var resultWorkbook = new XLWorkbook(resultsStream);
+
+            // Then
+            using (new AssertionScope())
+            {
+                foreach (var resultWorksheet in resultWorkbook.Worksheets)

```

Comment:
I believe this is a sort of one-way subset check? If the expectedWorkbook contains worksheets or data ranges that the resultWorkbook doesn't, then the test would still pass. Not sure if this is what you were referring to earlier? Maybe you check that the number of worksheets and number of cells match (then equivalence should follow).

Reply:
We also need to check ordering of the records is the same, but yeah I do think we should perhaps also be checking the number of sheets/filled cells etc to check we have the same. The intention was a complete equivalence check, but there is extra processing that happens with the excel file that means the byte arrays are not comparable, so we have to do some kind of cell by cell comparisons.